File: scripts/parity-runners/go-scenario/audio.go
package main

import (
	"encoding/json"
	"fmt"

	lkaudio "github.com/cavos-io/rtp-agent/core/audio"
	audiomodel "github.com/cavos-io/rtp-agent/core/audio/model"
)

func runAudioUtils(input json.RawMessage) (any, error) {
	var payload struct {
		Action string `json:"action"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.Action == "" {
		payload.Action = "silence_shape"
	}
	frameEvent := func(name string, frame *audiomodel.AudioFrame, original *audiomodel.AudioFrame) map[string]any {
		allZero := true
		for _, b := range frame.Data {
			if b != 0 {
				allZero = false
				break
			}
		}
		return map[string]any{
			"name":                name,
			"sample_rate":         frame.SampleRate,
			"num_channels":        frame.NumChannels,
			"samples_per_channel": frame.SamplesPerChannel,
			"data_bytes":          len(frame.Data),
			"all_zero":            allZero,
			"duration":            lkaudio.CalculateFrameDuration(frame),
			"same_object":         frame == original && original != nil,
		}
	}
	samples := func(frames []*audiomodel.AudioFrame) []uint32 {
		out := make([]uint32, 0, len(frames))
		for _, frame := range frames {
			out = append(out, frame.SamplesPerChannel)
		}
		return out
	}
	switch payload.Action {
	case "silence_shape":
		frame := lkaudio.SilenceFrame(0.02, 16000, 2)
		event := frameEvent("silence_shape", frame, nil)
		event["calculated_duration"] = lkaudio.CalculateAudioDuration([]*audiomodel.AudioFrame{frame})
		return map[string]any{"contract": "audio-utils", "events": []map[string]any{event}}, nil
	case "silence_zero":
		frame := lkaudio.SilenceFrame(0, 16000, 1)
		return map[string]any{"contract": "audio-utils", "events": []map[string]any{frameEvent("silence_zero", frame, nil)}}, nil
	case "silence_like":
		source := &audiomodel.AudioFrame{
			Data:              []byte{1, 2, 3, 4},
			SampleRate:        16000,
			NumChannels:       1,
			SamplesPerChannel: 2,
		}
		frame := lkaudio.SilenceFrameLike(source)
		return map[string]any{"contract": "audio-utils", "events": []map[string]any{frameEvent("silence_like", frame, source)}}, nil
	case "byte_stream_default":
		stream := lkaudio.NewAudioByteStream(16000, 1, 0)
		frames := stream.Push(make([]byte, 1600*2))
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "byte_stream_default", "frame_count": len(frames), "samples": samples(frames)},
			},
		}, nil
	case "byte_stream_write_alias":
		stream := lkaudio.NewAudioByteStream(16000, 1, 320)
		frames := stream.Write(make([]byte, 320*2))
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "byte_stream_write_alias", "frame_count": len(frames), "samples": samples(frames)},
			},
		}, nil
	case "byte_stream_progressive":
		stream := lkaudio.NewAudioByteStreamWithOptions(16000, 1, 3200, lkaudio.AudioByteStreamOptions{Progressive: true})
		frames := stream.Push(make([]byte, (320+640+1280)*2))
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "byte_stream_progressive", "frame_count": len(frames), "samples": samples(frames)},
			},
		}, nil
	case "byte_stream_flush_incomplete":
		stream := lkaudio.NewAudioByteStream(16000, 2, 1600)
		stream.Push([]byte{1, 2, 3})
		first := stream.Flush()
		stream.Push([]byte{4})
		second := stream.Flush()
		secondData := []byte{}
		if len(second) > 0 {
			secondData = second[0].Data
		}
		secondDataValues := make([]int, 0, len(secondData))
		for _, value := range secondData {
			secondDataValues = append(secondDataValues, int(value))
		}
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "byte_stream_flush_incomplete", "first_frame_count": len(first), "second_frame_count": len(second), "second_data": secondDataValues},
			},
		}, nil
	case "byte_stream_clear_progressive":
		stream := lkaudio.NewAudioByteStreamWithOptions(16000, 1, 3200, lkaudio.AudioByteStreamOptions{Progressive: true})
		stream.Push(make([]byte, 320*2))
		stream.Clear()
		frames := stream.Push(make([]byte, 320*2))
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "byte_stream_clear_progressive", "frame_count": len(frames), "samples": samples(frames)},
			},
		}, nil
	case "array_buffer_reject_oversized":
		buffer := lkaudio.NewAudioArrayBuffer(2, 16000)
		_, err := buffer.PushFrame(&audiomodel.AudioFrame{
			Data:              make([]byte, 6),
			SampleRate:        16000,
			NumChannels:       1,
			SamplesPerChannel: 3,
		})
		message := ""
		if err != nil {
			message = err.Error()
		}
		return map[string]any{
			"contract": "audio-utils",
			"events": []map[string]any{
				{"name": "array_buffer_reject_oversized", "error": err != nil, "error_message": message},
			},
		}, nil
	default:
		return nil, fmt.Errorf("unsupported audio utils action %q", payload.Action)
	}
}

File: scripts/parity-runners/python-scenarios/audio.py
from common import *  # noqa: F403

def audio_utils(input_data: Any) -> dict[str, Any]:
    action = input_data.get("action", "silence_shape")
    module = load_reference_audio()

    def frame_event(name: str, frame: Any, original: Any | None = None) -> dict[str, Any]:
        return {
            "name": name,
            "sample_rate": frame.sample_rate,
            "num_channels": frame.num_channels,
            "samples_per_channel": frame.samples_per_channel,
            "data_bytes": len(frame.data),
            "all_zero": all(byte == 0 for byte in frame.data),
            "duration": frame.duration,
            "same_object": frame is original if original is not None else False,
        }

    if action == "silence_shape":
        frame = module.silence_frame(0.02, 16000, 2)
        event = frame_event("silence_shape", frame)
        event["calculated_duration"] = module.calculate_audio_duration([frame])
        return {"contract": "audio-utils", "events": [event]}
    if action == "silence_zero":
        frame = module.silence_frame(0, 16000, 1)
        return {"contract": "audio-utils", "events": [frame_event("silence_zero", frame)]}
    if action == "silence_like":
        source = module.rtc.AudioFrame(
            data=bytes([1, 2, 3, 4]),
            sample_rate=16000,
            num_channels=1,
            samples_per_channel=2,
        )
        frame = module.silence_frame_like(source)
        return {"contract": "audio-utils", "events": [frame_event("silence_like", frame, source)]}
    if action == "byte_stream_default":
        stream = module.AudioByteStream(16000, 1)
        frames = stream.push(bytes(1600 * 2))
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "byte_stream_default",
                    "frame_count": len(frames),
                    "samples": [frame.samples_per_channel for frame in frames],
                }
            ],
        }
    if action == "byte_stream_write_alias":
        stream = module.AudioByteStream(16000, 1, 320)
        frames = stream.write(bytes(320 * 2))
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "byte_stream_write_alias",
                    "frame_count": len(frames),
                    "samples": [frame.samples_per_channel for frame in frames],
                }
            ],
        }
    if action == "byte_stream_progressive":
        stream = module.AudioByteStream(16000, 1, 3200, progressive=True)
        frames = stream.push(bytes((320 + 640 + 1280) * 2))
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "byte_stream_progressive",
                    "frame_count": len(frames),
                    "samples": [frame.samples_per_channel for frame in frames],
                }
            ],
        }
    if action == "byte_stream_flush_incomplete":
        stream = module.AudioByteStream(16000, 2, 1600)
        stream.push(bytes([1, 2, 3]))
        first = stream.flush()
        stream.push(bytes([4]))
        second = stream.flush()
        data = list(second[0].data) if second else []
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "byte_stream_flush_incomplete",
                    "first_frame_count": len(first),
                    "second_frame_count": len(second),
                    "second_data": data,
                }
            ],
        }
    if action == "byte_stream_clear_progressive":
        stream = module.AudioByteStream(16000, 1, 3200, progressive=True)
        stream.push(bytes(320 * 2))
        stream.clear()
        frames = stream.push(bytes(320 * 2))
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "byte_stream_clear_progressive",
                    "frame_count": len(frames),
                    "samples": [frame.samples_per_channel for frame in frames],
                }
            ],
        }
    if action == "array_buffer_reject_oversized":
        buffer = module.AudioArrayBuffer(buffer_size=2, sample_rate=16000)
        frame = module.rtc.AudioFrame(
            data=bytes(6),
            sample_rate=16000,
            num_channels=1,
            samples_per_channel=3,
        )
        error = False
        message = ""
        try:
            buffer.push_frame(frame)
        except Exception as exc:
            error = True
            message = str(exc)
        return {
            "contract": "audio-utils",
            "events": [
                {
                    "name": "array_buffer_reject_oversized",
                    "error": error,
                    "error_message": message,
                }
            ],
        }
    raise ValueError(f"unsupported audio utils action {action!r}")


